src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs

```diff
@@ -663,11 +663,100 @@ public void AnalyzeCommandBase_ReportsWarningOnUnsupportedPlatformForRuleAndNoRu
+
+                IList<Notification> executionNotifications = run.Invocations[0].ToolExecutionNotifications;
+
+                switch (trace)
+                {
+                    case DefaultTraces.None:
+                    {
+                        if (executionNotificationsCount > 0 || configurationNotificationCount > 0)
+                        {
+                            sb.AppendLine($"\t{trace} : observed notifications when tracing was disabled.");
+                        }
+                        break;
+                    }
+                    case DefaultTraces.ScanTime:
+                    {
+                        // There is only one end-to-end scan time notification.
+                        if (executionNotifications?.Count != 1)
+                        {
+                            sb.AppendLine($"\t{trace} : expected 1 notification but saw {executionNotifications?.Count ?? 0}.");
+                            continue;
+                        }
+
+                        if (executionNotifications?.Where(t => t.Message.Text.Contains("elapsed")).Count() != 1)
+                        {
+                            sb.AppendLine($"\t{trace} : did not observe term 'elapsed' in scan timing notifications.");
+                        }
+                        break;
+                    }
+                    case DefaultTraces.RuleScanTime:
+                    {
+                        // We expect every rule to generate timing data for every applicable scan target.
+                        int rulesCount = run.Tool.Driver.Rules.Count;
+                        int validTargetsCount = ALL_COUNT - NOT_APPLICABLE_COUNT;
+                        int expectedNotificationsCount = rulesCount * validTargetsCount;
+
+                        // We expected timing data for every rule.
+                        if (executionNotificationsCount != expectedNotificationsCount)
+                        {
+                            sb.AppendLine($"\t{trace} : expected {expectedNotificationsCount} notifications but saw {executionNotificationsCount}.");
+                            continue;
+                        }
+
+                        if (executionNotifications?.Where(t => t.Message.Text.Contains("elapsed")).Count() != expectedNotificationsCount)
+                        {
+                            sb.AppendLine($"\t{trace} : did not observe term 'elapsed' in rule timing notifications.");
+                        }
+
+                        if (executionNotifications?.GroupBy(t => t.AssociatedRule.Id).Count() != rulesCount)
+                        {
+                            sb.AppendLine($"\t{trace} : did not observe timing notifications for every rule.");
+                        }
+
+                        break;
+                    }
+                }
+            }
+            sb.Length.Should().Be(0, $"test cases failed : {Environment.NewLine}{sb}");
```

Comment:
> $"test cases failed : {Environment.NewLine}{sb}" [](http://example.com/codeflow?start=36&length=49)

notice this construct which maximizes readability in the VS test window. this reporting subtlety should be captured in some test helpers sometime soon.